src/codegen/operator/update_translator.cpp

```diff
@@ -0,0 +1,157 @@
+                                   Pipeline &pipeline)
+    : OperatorTranslator(context, pipeline), update_plan_(update_plan),
+      table_storage_(*update_plan.GetTable()->GetSchema()) {
+  // Create the translator for our child and derived attributes
+  context.Prepare(*update_plan.GetChild(0), pipeline);
+
+  const auto *project_info = update_plan_.GetProjectInfo();
+  auto target_list = project_info->GetTargetList();
+  for (const auto &target : target_list) {
+    const auto &derived_attribute = target.second;
+    context.Prepare(*derived_attribute.expr);
+  }
+  // Prepare for updater
+  updater_state_id_ = context.GetRuntimeState().RegisterState("updater",
+      UpdaterProxy::GetType(GetCodeGen()));
+}
+
+bool IsTarget(const TargetList &target_list, uint32_t index) {
+  for (const auto &target : target_list) {
+    if (target.first == index)
+      return true;
+  }
+  return false;
+}
+
+void UpdateTranslator::InitializeState() {
+  auto &codegen = GetCodeGen();
+  auto &context = GetCompilationContext();
+
+  // Prepare for all the information to be handed over to the updater
+  // Get the transaction pointer
+  llvm::Value *txn_ptr = context.GetTransactionPtr();
+
+  // Get the table object pointer
+  storage::DataTable *table = update_plan_.GetTable();
+  llvm::Value *table_ptr = codegen.Call(StorageManagerProxy::GetTableWithOid,
+      {GetCatalogPtr(), codegen.Const32(table->GetDatabaseOid()),
+       codegen.Const32(table->GetOid())});
+
+  // Get the target list's raw vectors and their sizes
+  // : this is required when installing a new version at updater
+  const auto *project_info = update_plan_.GetProjectInfo();
+  llvm::Value *target_vector_ptr = codegen->CreateIntToPtr(
+      codegen.Const64((int64_t)project_info->GetTargetList().data()),
+      TargetProxy::GetType(codegen)->getPointerTo());
+  llvm::Value *target_vector_size_ptr =
+      codegen.Const32((int32_t)project_info->GetTargetList().size());
+
+  // Initialize the inserter with txn and table
+  llvm::Value *updater = LoadStatePtr(updater_state_id_);
+  codegen.Call(UpdaterProxy::Init, {updater, txn_ptr, table_ptr,
+                                    target_vector_ptr, target_vector_size_ptr});
+}
+
+void UpdateTranslator::Produce() const {
+  GetCompilationContext().Produce(*update_plan_.GetChild(0));
+}
+
+void UpdateTranslator::Consume(ConsumerContext &, RowBatch::Row &row) const {
+  auto &codegen = GetCodeGen();
+  auto &context = GetCompilationContext();
+  const auto *project_info = update_plan_.GetProjectInfo();
+  auto target_list = project_info->GetTargetList();
+  auto direct_map_list = project_info->GetDirectMapList();
+  uint32_t column_num =
+      static_cast<uint32_t>(target_list.size() + direct_map_list.size());
+  auto *scan =
+      static_cast<const planner::AbstractScan *>(update_plan_.GetChild(0));
+  std::vector<const planner::AttributeInfo *> ais;
+  scan->GetAttributes(ais);
```

Comment:
Shouldn't this be done in `UpdatePlan`? At this point, the plans should already know all attribute information.